app/eth2wrap/synthproposer.go

```diff
@@ -0,0 +1,205 @@
+	vals, err := eth2Cl.ValidatorsByPubKey(ctx, fmt.Sprint(epochSlot), c.pubkeys)
+	if err != nil {
+		return nil, err
+	}
+
+	var activeIdxs []eth2p0.ValidatorIndex
+	for _, val := range vals {
+		if !val.Status.IsActive() {
+			continue
+		}
+		activeIdxs = append(activeIdxs, val.Index)
+	}
+
+	// Get actual duties for all validators for the epoch.
+	duties, err = eth2Cl.ProposerDuties(ctx, epoch, activeIdxs)
+	if err != nil {
+		return nil, err
+	}
+
+	// Mark those not requiring synthetic duties.
+	noSynth := make(map[eth2p0.ValidatorIndex]bool)
+	for _, duty := range duties {
+		noSynth[duty.ValidatorIndex] = true
+	}
+
+	// Deterministic synthetic duties for the rest.
+	synthSlots := make(map[eth2p0.Slot]bool)
+	for _, idx := range c.shuffleFunc(epoch, activeIdxs) {
+		if noSynth[idx] {
+			continue
+		}
+
+		synthSlot := epochSlot + eth2p0.Slot(idx)%eth2p0.Slot(slotsPerEpoch)
+		if _, ok := synthSlots[synthSlot]; ok {
+			// We already have a synth proposer for this slot.
+			continue
+		}
+
+		synthSlots[synthSlot] = true
+		duties = append(duties, &eth2v1.ProposerDuty{
+			PubKey:         vals[idx].Validator.PublicKey,
+			Slot:           synthSlot,
+			ValidatorIndex: idx,
+		})
+	}
+
+	// Cache the values for the epoch
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.fifo = append(c.fifo, epoch)
+	c.duties[epoch] = duties
+	c.synths[epoch] = synthSlots
+
+	// Trim the cache
+	if len(c.fifo) > fifoEpochMax {
+		delete(c.duties, c.fifo[0])
+		delete(c.synths, c.fifo[0])
+		c.fifo = c.fifo[1:]
+	}
+
+	return duties, nil
+}
+
+// IsSynthetic returns true if the slot is a synthetic proposer duty from via a read-through cache.
+func (c *synthProposerCache) IsSynthetic(ctx context.Context, eth2Cl synthProposerEth2Provider, slot eth2p0.Slot) (bool, error) {
+	// Get the epoch.
+	slotsPerEpoch, err := eth2Cl.SlotsPerEpoch(ctx)
+	if err != nil {
+		return false, err
+	}
+	epoch := eth2p0.Epoch(slot) / eth2p0.Epoch(slotsPerEpoch)
+
+	// Ensure that cache is populated.
+	_, err = c.Duties(ctx, eth2Cl, epoch)
+	if err != nil {
+		return false, err
+	}
+
+	// Return the result.
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	return c.synths[epoch][slot], nil
+}
+
+// eth2Shuffle is the eth2 pseudo-random (deterministic) shuffle function.
+func eth2Shuffle(epoch eth2p0.Epoch, indices []eth2p0.ValidatorIndex) []eth2p0.ValidatorIndex {
+	var uints []uint64
+	for _, i := range indices {
+		uints = append(uints, uint64(i))
+	}
+
+	// Use little endian epoch as the seed.
+	var seed [32]byte
+	binary.LittleEndian.PutUint64(seed[:], uint64(epoch))
+
+	const rounds = 90 // From https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#misc-1
+
+	shuffle.ShuffleList(getStandardHashFn(), uints, rounds, seed)
+
+	var resp []eth2p0.ValidatorIndex
+	for _, i := range uints {
+		resp = append(resp, eth2p0.ValidatorIndex(i))
+	}
+
+	return resp
+}
+
+// getStandardHashFn returns a standard sha256 hash function as per eth2.
```

Comment:
discussion: is the hash function mentioned somewhere in the specs?

Reply:
go this from go-eth2-shuffle.

Reply:
okay, can add reference as comment: https://github.com/protolambda/eth2-shuffle/blob/master/test_util.go